File: compress/src/py/compresslib/bitio.py
# -*- coding: utf-8 -*-
"""비트 writer/reader — SPEC §1.

이 덱에서 두 가지 비트 순서를 다 쓴다. 우리가 만드는 형식은 전부
MSB 먼저이고, DEFLATE 만 LSB 먼저다. 둘은 바꿔 쓸 수 없다.

가장 자주 갈라지는 자리는 flush 의 채움이다. **채움은 0 비트**이고,
쌓인 비트가 없으면 바이트를 내보내지 않는다. 1로 채우는 구현도
자기들끼리는 왕복이 되니, 골든 벡터를 맞춰 보기 전에는 아무도 모른다.

읽기·쓰기 모두 비트 하나당 O(1).
"""
from compresslib import varint


class MsbWriter:
    """첫 비트가 첫 바이트의 7번 비트에 앉는 writer."""

    def __init__(self):
        self._out = bytearray()
        self._buf = 0
        self._n = 0

    def write_bit(self, bit):
        self._buf |= (bit & 1) << (7 - self._n)
        self._n += 1
        if self._n == 8:
            self._out.append(self._buf)
            self._buf = 0
            self._n = 0

    def write_bits(self, value, count):
        """값의 낮은 count 비트를 높은 쪽부터 쓴다."""
        for i in range(count - 1, -1, -1):
            self.write_bit((value >> i) & 1)

    # MSB 먼저인 스트림에서는 값도 부호도 같은 순서다. 이름만 따로 둔
    # 것은 LsbWriter 와 쓰는 쪽 코드를 똑같이 만들기 위해서다.
    write_code = write_bits

    def flush(self):
        if self._n:
            self._out.append(self._buf)
            self._buf = 0
            self._n = 0

    def bytes(self):
        return bytes(self._out)


class MsbReader:
    """MsbWriter 가 쓴 것을 그대로 되읽는다."""

    def __init__(self, src, pos=0):
        self.src = src
        self.pos = pos
        self._buf = 0
        self._n = 0

    def read_bit(self):
        if self._n == 0:
            if self.pos >= len(self.src):
                raise ValueError('비트 스트림이 바닥났다')
            self._buf = self.src[self.pos]
            self.pos += 1
            self._n = 8
        self._n -= 1
        return (self._buf >> self._n) & 1

    def read_bits(self, count):
        value = 0
        for _ in range(count):
            value = (value << 1) | self.read_bit()
        return value

    def align(self):
        """바이트 경계까지 버린다 (DEFLATE 의 stored 블록에서 쓴다)."""
        self._n = 0


class LsbWriter:
    """첫 비트가 첫 바이트의 0번 비트에 앉는 writer — DEFLATE 용."""

    def __init__(self):
        self._out = bytearray()
        self._buf = 0
        self._n = 0

    def write_bit(self, bit):
        self._buf |= (bit & 1) << self._n
        self._n += 1
        if self._n == 8:
            self._out.append(self._buf)
            self._buf = 0
            self._n = 0

    def write_bits(self, value, count):
        """헤더 칸·여분 비트 — 낮은 비트부터."""
        for i in range(count):
            self.write_bit((value >> i) & 1)

    def write_code(self, code, count):
        """허프만 부호 — 같은 LSB 스트림에 높은 비트부터 넣는다.

        RFC 1951 의 가장 헷갈리는 한 줄이다. 스트림은 LSB 먼저인데
        부호만 MSB 먼저다. 두 함수로 갈라 두면 부르는 쪽이 안 헷갈린다.
        """
        for i in range(count - 1, -1, -1):
            self.write_bit((code >> i) & 1)

    def flush(self):
        if self._n:
            self._out.append(self._buf)
            self._buf = 0
            self._n = 0

    def align(self):
        """바이트 경계까지 0으로 채운다 — stored 블록 앞에서."""
        self.flush()

    def bytes(self):
        return bytes(self._out)


class LsbReader:

    def __init__(self, src, pos=0):
        self.src = src
        self.pos = pos
        self._buf = 0
        self._n = 0

    def read_bit(self):
        if self._n == 0:
            if self.pos >= len(self.src):
                raise ValueError('비트 스트림이 바닥났다')
            self._buf = self.src[self.pos]
            self.pos += 1
            self._n = 8
        bit = self._buf & 1
        self._buf >>= 1
        self._n -= 1
        return bit

    def read_bits(self, count):
        value = 0
        for i in range(count):
            value |= self.read_bit() << i
        return value

    def align(self):
        self._n = 0


# ------------------------------------------------------------ 골든 코덱
# SPEC §1.4. 압축기가 아니라 "비트 writer 가 진짜로 비트 단위로 도는가"
# 를 골든 벡터로 묶어 두기 위한 껍데기다. 앞의 0비트 셋이 요점 — 모든
# 바이트를 바이트 경계 밖으로 밀어내므로, 몰래 memcpy 하는 구현은 다른
# 파일을 낸다.
PAD_BITS = 3


def encode(src):
    w = MsbWriter()
    w.write_bits(0, PAD_BITS)
    for b in src:
        w.write_bits(b, 8)
    w.flush()
    return varint.put(len(src)) + w.bytes()


def decode(src):
    n, pos = varint.get_length(src)
    r = MsbReader(src, pos)
    r.read_bits(PAD_BITS)
    return bytes(r.read_bits(8) for _ in range(n))

File: compress/src/java/compresslib/BitIO.java
package compresslib;

/**
 * 비트 writer/reader — SPEC §1.
 *
 * <p>우리 형식은 전부 MSB 먼저이고 DEFLATE 만 LSB 먼저다. 가장 자주
 * 갈라지는 자리는 flush 의 채움이다 — <b>채움은 0</b> 이고, 쌓인 비트가
 * 없으면 바이트를 내보내지 않는다.
 */
public final class BitIO {
  private BitIO() {}

  /** 비트를 하나씩 내주는 것. 허프만 복호기가 이 모양만 요구한다. */
  public interface BitSource {
    int readBit();
  }

  public static final class MsbWriter {
    private final ByteBuf out = new ByteBuf();
    private int buf;
    private int n;

    public void writeBit(int bit) {
      buf |= (bit & 1) << (7 - n);
      if (++n == 8) {
        out.push(buf);
        buf = 0;
        n = 0;
      }
    }

    public void writeBits(long v, int count) {
      for (int i = count - 1; i >= 0; i--) {
        writeBit((int) ((v >>> i) & 1));
      }
    }

    /**
     * MSB 스트림에서는 값도 부호도 같은 순서다. 이름만 LsbWriter 와
     * 맞춘다.
     */
    public void writeCode(long code, int count) {
      writeBits(code, count);
    }

    public int bitPos() {
      return out.size() * 8 + n;
    }

    public void flush() {
      if (n > 0) {
        out.push(buf);
        buf = 0;
        n = 0;
      }
    }

    public byte[] bytes() {
      return out.bytes();
    }
  }

  public static final class MsbReader implements BitSource {
    private final byte[] src;
    public int pos;
    private int buf;
    private int n;

    public MsbReader(byte[] src, int pos) {
      this.src = src;
      this.pos = pos;
    }

    @Override
    public int readBit() {
      if (n == 0) {
        if (pos >= src.length) {
          throw new CodecException("비트 스트림이 바닥났다");
        }
        buf = src[pos++] & 0xFF;
        n = 8;
      }
      n--;
      return (buf >>> n) & 1;
    }

    public long readBits(int count) {
      long v = 0;
      for (int i = 0; i < count; i++) {
        v = (v << 1) | readBit();
      }
      return v;
    }

    public void align() {
      n = 0;
    }
  }

  public static final class LsbWriter {
    private final ByteBuf out = new ByteBuf();
    private int buf;
    private int n;

    public void writeBit(int bit) {
      buf |= (bit & 1) << n;
      if (++n == 8) {
        out.push(buf);
        buf = 0;
        n = 0;
      }
    }

    public void writeBits(long v, int count) {
      for (int i = 0; i < count; i++) {
        writeBit((int) ((v >>> i) & 1));
      }
    }

    /**
     * 허프만 부호만 같은 LSB 스트림에 높은 비트부터 넣는다 (RFC 1951).
     */
    public void writeCode(long code, int count) {
      for (int i = count - 1; i >= 0; i--) {
        writeBit((int) ((code >>> i) & 1));
      }
    }

    public int bitPos() {
      return out.size() * 8 + n;
    }

    public void flush() {
      if (n > 0) {
        out.push(buf);
        buf = 0;
        n = 0;
      }
    }

    public void align() {
      flush();
    }

    public byte[] bytes() {
      return out.bytes();
    }
  }

  public static final class LsbReader implements BitSource {
    private final byte[] src;
    public int pos;
    private int buf;
    private int n;

    public LsbReader(byte[] src, int pos) {
      this.src = src;
      this.pos = pos;
    }

    @Override
    public int readBit() {
      if (n == 0) {
        if (pos >= src.length) {
          throw new CodecException("비트 스트림이 바닥났다");
        }
        buf = src[pos++] & 0xFF;
        n = 8;
      }
      int bit = buf & 1;
      buf >>>= 1;
      n--;
      return bit;
    }

    public long readBits(int count) {
      long v = 0;
      for (int i = 0; i < count; i++) {
        v |= ((long) readBit()) << i;
      }
      return v;
    }

    public void align() {
      n = 0;
    }
  }

  // 골든 코덱 (SPEC §1.4). 앞의 0비트 셋이 요점 — 모든 바이트를 바이트
  // 경계 밖으로 밀어내므로, 몰래 복사하는 구현은 다른 파일을 낸다.
  public static final int PAD_BITS = 3;

  public static byte[] encode(byte[] src) {
    MsbWriter w = new MsbWriter();
    w.writeBits(0, PAD_BITS);
    for (byte b : src) {
      w.writeBits(b & 0xFF, 8);
    }
    w.flush();
    ByteBuf out = new ByteBuf();
    Varint.put(out, src.length);
    out.extend(w.bytes());
    return out.bytes();
  }

  public static byte[] decode(byte[] src) {
    int[] h = Varint.getLength(src, 0);
    MsbReader r = new MsbReader(src, h[1]);
    r.readBits(PAD_BITS);
    byte[] out = new byte[h[0]];
    for (int i = 0; i < out.length; i++) {
      out[i] = (byte) r.readBits(8);
    }
    return out;
  }
}
